mindmeld/query_cache.py

```diff
@@ -14,131 +14,201 @@
 """
 This module contains the query cache implementation.
 """
+from functools import lru_cache
+from hashlib import sha256
+import json
 import logging
 import os
-import shutil
+import sqlite3
 
-from sklearn.externals import joblib
-
-from ._version import get_mm_version
-from .path import GEN_FOLDER, QUERY_CACHE_PATH, QUERY_CACHE_TMP_PATH
+from .path import GEN_FOLDER, QUERY_CACHE_DB_PATH
+from .core import ProcessedQuery
 
 logger = logging.getLogger(__name__)
 
 
 class QueryCache:
-    """
-    An object that stores ProcessedQuery objects in memory to save time on reloading.
-    ProcessedQuery objects consist of the query itself, the domain/intent classifications,
-    recognized entities in the query, and more.
-    """
+    '''
+    QueryCache stores ProcessedQuerys and associated metadata in a sqlite3 backed
+    cache to save processing time on reloading the examples later.
+    '''
 
     def __init__(self, app_path):
-        self.app_path = app_path
-        self.is_dirty = False
-        # We initialize cached_queries to None instead of {} since
-        # we want to lazy load it from disk only when necessary ie during
-        # set, get and dump ops. This allows us to run the application
-        # faster.
-        self._cached_queries = None
-        self.gen_folder = GEN_FOLDER.format(app_path=self.app_path)
-        self.main_cache_location = QUERY_CACHE_PATH.format(app_path=self.app_path)
-        self.tmp_cache_location = QUERY_CACHE_TMP_PATH.format(app_path=self.app_path)
-
-    @property
-    def cached_queries(self):
-        """A dictionary containing all the cached queries"""
-        if self._cached_queries is None:
-            self.load()
-
-        return self._cached_queries
-
-    @property
-    def versioned_data(self):
-        """A dictionary containing the MindMeld version in addition to any cached queries."""
-        return {"mm_version": get_mm_version(), "cached_queries": self.cached_queries}
-
-    def set_value(self, domain, intent, query_text, processed_query):
-        """
-        Set value for the corresponding argument parameters.
+        # make generated directory if necessary
+        gen_folder = GEN_FOLDER.format(app_path=app_path)
+        if not os.path.isdir(gen_folder):
+            os.makedirs(gen_folder)
+
+        db_file_location = QUERY_CACHE_DB_PATH.format(app_path=app_path)
+        self.connection = sqlite3.connect(db_file_location)
+        cursor = self.connection.cursor()
+
+        if not self.compatible_version():
+            cursor.execute('''
+            DROP TABLE IF EXISTS queries;
+            ''')
+            cursor.execute('''
+            DROP TABLE IF EXISTS version;
+            ''')
+        # Create table to store queries
+        cursor.execute('''
+        CREATE TABLE IF NOT EXISTS queries
+        (hash_id TEXT PRIMARY KEY, query TEXT, raw_query TEXT, domain TEXT, intent TEXT)
```

Comment:
I propose also adding a composite primary key containing app_path and text as the primary key. Since we have users who use MindMeld in a multi-application mode on the same python process, we need to be able to distinguish same text queries which are in different domains/intents in the different applications. So the app_path + text would be a unique composite key in such a case.

Reply:
I'm not sure I understand the problem.  The `hash_id` primary key is derived from domain, intent, and raw_text.  The only time a conflict would occur is if two apps have the same example in the same domain and intent.  I expect that case would actually function correctly as it is now?